src/main/java/filter/SemanticException.java

```diff
@@ -0,0 +1,17 @@
+package filter;
+
+import filter.expression.QualifierType;
+
+/**
+ * Thrown to indicate an invalid input has been provided to a Qualifier
```

Comment:
This is a little specific. It can indicate other sorts of failures when trying to interpret these expressions. Maybe write something along those lines (problems with interpretation), and give a few examples? "For example, invalid qualifier inputs, multiple qualifiers which should only appear once, etc."